### ISSUE TITLE
[JENKINS-44423] Upgrading blueocean to plugin-pom:2.8+ breaks unit tests

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4141]
<p>Vivek and I have been able to reliably reproduce test failures in blueocean-plugin when using the plugin-pom:2.28. I also saw these errors using a 2.27 snapshot deployed in support of <a href="https://issues.jenkins.io/browse/JENKINS-43411" title="Evaluate yarn" class="issue-link" data-issue-key="JENKINS-43411"><del>JENKINS-43411</del></a></p>

<p>The tests pass when run in CI, or when individual tests are run locally. But when running a full build locally (mvn clean install) the tests fail 100% of the time. Vivek believes it's related to jenkins-test-harness:2.22.</p>

<p>blueocean-pipeline-api-impl has the following broken tests:</p>

<p>Failed tests: <br/>
  BranchContainerImplTest.testBranchOrdering:57 expected:<200> but was:<403><br/>
  MultiBranchTest.createUserFavouriteMultibranchBranchTest:666 expected:<200> but was:<403><br/>
  MultiBranchTest.createUserFavouriteMultibranchTopLevelTest:591 expected:<200> but was:<403><br/>
Tests in error: <br/>
  CredentialApiTest.createSshCredentialUsingDefaultSshOnMasterInUserStore:250 » Runtime<br/>
  CredentialApiTest.createSshCredentialUsingDirectSshInUserStore:213 » Runtime c...<br/>
  CredentialApiTest.createUsingUsernamePasswordInUserStore:186 » Runtime com.mas...<br/>
  MultiBranchTest.favoritedFromClassicTest:741 » Runtime com.mashape.unirest.htt...</p>

<p>And here's one of the more interesting error details:</p>

<pre>
{
  <span class="code-quote">"message"</span> : <span class="code-quote">"This user <span class="code-quote">'anonymous'</span> cannot access resource owned by <span class="code-quote">'alice'</span>"</span>,
  <span class="code-quote">"code"</span> : 403,
  <span class="code-quote">"errors"</span> : [ ]
}; line: 1, column: 1]
	at io.jenkins.blueocean.<span class="code-keyword">rest</span>.impl.pipeline.PipelineBaseTest$RequestBuilder.build(PipelineBaseTest.java:542)
	at io.jenkins.blueocean.<span class="code-keyword">rest</span>.impl.pipeline.MultiBranchTest.favoritedFromClassicTest(MultiBranchTest.java:741)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.jvnet.hudson.test.JenkinsRule$2.evaluate(JenkinsRule.java:547)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers">cliffmeyers</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423" target="_blank">Upgrading blueocean to plugin-pom:2.8+ breaks unit tests</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: technical-debt
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-09-22T03:49:39+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Vivek and I have been able to reliably reproduce test failures in blueocean-plugin when using the plugin-pom:2.28. I also saw these errors using a 2.27 snapshot deployed in support of <a href="https://issues.jenkins.io/browse/JENKINS-43411" title="Evaluate yarn" class="issue-link" data-issue-key="JENKINS-43411"><del>JENKINS-43411</del></a></p>

<p>The tests pass when run in CI, or when individual tests are run locally. But when running a full build locally (mvn clean install) the tests fail 100% of the time. Vivek believes it's related to jenkins-test-harness:2.22.</p>

<p>blueocean-pipeline-api-impl has the following broken tests:</p>

<p>Failed tests: 
  BranchContainerImplTest.testBranchOrdering:57 expected:&lt;200&gt; but was:&lt;403&gt;
  MultiBranchTest.createUserFavouriteMultibranchBranchTest:666 expected:&lt;200&gt; but was:&lt;403&gt;
  MultiBranchTest.createUserFavouriteMultibranchTopLevelTest:591 expected:&lt;200&gt; but was:&lt;403&gt;
Tests in error: 
  CredentialApiTest.createSshCredentialUsingDefaultSshOnMasterInUserStore:250 » Runtime
  CredentialApiTest.createSshCredentialUsingDirectSshInUserStore:213 » Runtime c...
  CredentialApiTest.createUsingUsernamePasswordInUserStore:186 » Runtime com.mas...
  MultiBranchTest.favoritedFromClassicTest:741 » Runtime com.mashape.unirest.htt...</p>

<p>And here's one of the more interesting error details:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">{
  <span class="code-quote">"message"</span> : <span class="code-quote">"This user <span class="code-quote">'anonymous'</span> cannot access resource owned by <span class="code-quote">'alice'</span>"</span>,
  <span class="code-quote">"code"</span> : 403,
  <span class="code-quote">"errors"</span> : [ ]
}; line: 1, column: 1]
	at io.jenkins.blueocean.<span class="code-keyword">rest</span>.impl.pipeline.PipelineBaseTest$RequestBuilder.build(PipelineBaseTest.java:542)
	at io.jenkins.blueocean.<span class="code-keyword">rest</span>.impl.pipeline.MultiBranchTest.favoritedFromClassicTest(MultiBranchTest.java:741)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.jvnet.hudson.test.JenkinsRule$2.evaluate(JenkinsRule.java:547)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
</pre>
</div></div></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=300437&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300437">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> had spent some time investigating but had to shelve the work.</p>

<p>Comment on related PR: <a href="https://github.com/jenkinsci/blueocean-plugin/pull/981#issuecomment-302723928" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/981#issuecomment-302723928</a></p>

<p>"(Possibly just that you depend on Jenkins 2.x and you were assuming that detached plugins would be loaded implicitly in tests. This is no longer true; there is a TestPluginManager method to explicitly load them, but much better to have all plugins you expect to test against be test-scoped dependencies.)"</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> had spent some time investigating but had to shelve the work.</p>

<p>Comment on related PR: <a href="https://github.com/jenkinsci/blueocean-plugin/pull/981#issuecomment-302723928" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/981#issuecomment-302723928</a></p>

<p>"(Possibly just that you depend on Jenkins 2.x and you were assuming that detached plugins would be loaded implicitly in tests. This is no longer true; there is a TestPluginManager method to explicitly load them, but much better to have all plugins you expect to test against be test-scoped dependencies.)"</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=300493&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300493">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> so this needs some of <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a>'s time to unblock <a href="https://issues.jenkins.io/browse/JENKINS-43411" title="Evaluate yarn" class="issue-link" data-issue-key="JENKINS-43411"><del>JENKINS-43411</del></a>? cc <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> so this needs some of <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a>'s time to unblock <a href="https://issues.jenkins.io/browse/JENKINS-43411" title="Evaluate yarn" class="issue-link" data-issue-key="JENKINS-43411"><del>JENKINS-43411</del></a>? cc <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=300497&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300497">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>When do we need to upgrade that pom? that determines when vivek can chip away at this (likely have to juggled dependencies). </p>
</pre>
</details>

</ul>
</details>

<p>When do we need to upgrade that pom? that determines when vivek can chip away at this (likely have to juggled dependencies). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=300566&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300566">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It would be required in order to port the build to yarn. However I'd like the new extensibility system to land first because I think it will obviate the need for a lot of npm publishing, and there is a known bug (w/ slightly tedious workaround) when publishing to npm using yarn with a custom registry.</p>
</pre>
</details>

</ul>
</details>

<p>It would be required in order to port the build to yarn. However I'd like the new extensibility system to land first because I think it will obviate the need for a lot of npm publishing, and there is a known bug (w/ slightly tedious workaround) when publishing to npm using yarn with a custom registry.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=300633&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300633">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> right, on top of that, it sounds like npm is upping its game in response to this stuff... so lets keep this one parked. I think will need vivek to take a look at this to resolve it at some point right? </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> right, on top of that, it sounds like npm is upping its game in response to this stuff... so lets keep this one parked. I think will need vivek to take a look at this to resolve it at some point right? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=300719&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300719">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Right, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> is probably best suited. This is just a little too far outside my area of expertise to productively troubleshoot. Presumably we'll want to upgrade our parent POM at some point for reasons other than yarn, so that's when I opened this ticket so we don't forget it's an issue.</p>
</pre>
</details>

</ul>
</details>

<p>Right, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> is probably best suited. This is just a little too far outside my area of expertise to productively troubleshoot. Presumably we'll want to upgrade our parent POM at some point for reasons other than yarn, so that's when I opened this ticket so we don't forget it's an issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44423?focusedId=314559&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-314559">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> can we close this now as PCT work done? </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> can we close this now as PCT work done? </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-43411%22">JENKINS-43411</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35759%22">JENKINS-35759</a>]</i>
